presto-main/src/main/java/com/facebook/presto/execution/scheduler/CTEMaterializationTracker.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution.scheduler;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * Tracks the completion status of table-finish nodes that write temporary tables for CTE materialization.
+ * CTEMaterializationTracker manages a map of materialized CTEs and their associated materialization futures.
+ * When a stage includes a CTE table finish, it marks the corresponding CTE as materialized and completes
+ * the associated future. This signals the scheduler that some dependency has been resolved, prompting it
+ * to resume scheduling. The scheduler can call the clearAllFutures() method once it starts scheduling again
+ */
+public class CTEMaterializationTracker
+{
+    private final Map<String, SettableFuture<Void>> materializationFutures = new HashMap<>();
+
+    private final Map<String, Boolean> materializedCtes = new HashMap<>();
+
+    public synchronized ListenableFuture<Void> getFutureForCTE(String cteName)
```

Comment:
it would be nice to avoid all this locking. it seems like right now it's only needed to make the materializationFutures and materializedCTE calls atomic (because aside from that we still make assumptions about when cancel is or is not called, so it's not providing any thread safety with regard to the cancellation calls).

  I'm thinking it would be better to use a single ConcurrentHashMap<>, and then not clear the futures before cancellation and instead check in getFutureForCTE() that it hasn't been cancelled, and if it has create a new one (like the code you had before).  something like the following.

```
private final Map<String, SettableFuture<Void>> materializationFutures = new ConcurrentHashMap<>();

public ListenableFuture<Void> getFutureForCTE(String cteName)
    {
         return materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.iscancelled()) {
                return SettableFuture.create();
            }
            return existingFuture;
        });

   public void markCTEAsMaterialized(String cteName)
    {
         materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.iscancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
            existingFuture.set(null); // notify all listeners
            return existingFuture;
        });


```

Reply:
Not happy with the assumptions that we are making, will brainstorm for a better way but it won't be trivial

Reply:
```
public void markCTEAsMaterialized(String cteName)
    {
        materializationFutures.compute(cteName, (key, existingFuture) -> {
            if (existingFuture == null || existingFuture.isCancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
            existingFuture.set(null); // notify all listeners
            return existingFuture;
        });
    }
```
Also here completing the future when its cancelled
```
  if (existingFuture == null || existingFuture.isCancelled()) {
                SettableFuture completedFuture = SettableFuture.create();
                completedFuture.set(null);
                return completedFuture;
            }
```
is not really useful since there will be no listeners on the new future. the entire goal of the markCTEAsMaterialized() is to unblock the stages which were waiting on this future

Reply:
the point of setting a new completed future if the original one was canceled is in order to add it to the map. that way when someone calls getFutureForCTE() they will get a future that is done.

Reply:
Got it, first race condition is fixed.  For the second condition

> For part 2 It's also possible for the future to be canceled after it's returned, but before the consumer checks it or even before it's returned, but after this argument check

If we see the FixedSourcePartitionScheduler code 
```
 // If any CTE is not materialized, return a blocked ScheduleResult
            if (!blocked.isEmpty()) {
                return ScheduleResult.blocked(
                        false,
                        newTasks,
                        whenAnyComplete(blocked),
                        BlockedReason.WAITING_FOR_CTE_MATERIALIZATION,
                        0);
            }
```
Actually I don't think we are even cancelling the future in the CteMaterializationTracker. We are cancelling the future  which was wrapped by the `whenAnyComplete` so that problem seems non exisitent

Reply:
Okay I think I found a good fix. If we check the [FixedSourcePatitionScheduler](https://github.com/prestodb/presto/pull/24108/files#diff-1090adf62744bbb6103f894606a0574e2d8b040779e2279e9846a820c2d6a118R200) code, we can see that it actually returns a wrapped future in whenAnyComplete. This wrapped future which can be cancelled. I also debugged and verified that if this is cancelled the inner ones are also cancelled (maybe because we cancelled with mayInterruptIfRunning?) 
However if we make sure that the ctematerialization tracker always returns a `Futures.nonCancellationPropagating` then the future in the tracker can never be cancelled so in a way we have handled cancellations. 

```
if (!blocked.isEmpty()) {
                return ScheduleResult.blocked(
                        false,
                        newTasks,
                        whenAnyComplete(blocked),
                        BlockedReason.WAITING_FOR_CTE_MATERIALIZATION,
                        0);
            }
```